src/main/java/io/nats/client/impl/Headers.java

```diff
@@ -0,0 +1,274 @@
+	 * any existing values are removed and replaced with the new list.
+	 * null values are ignored. If all values are null, the put is ignored
+	 *
+	 * @param key the key
+	 * @param values the values
+	 * @throws IllegalArgumentException if the key is null or empty or contains invalid characters
+	 *         -or- if any value contains invalid characters
+	 */
+	public Headers put(String key, String... values) {
+		if (values != null) {
+			_put(key, Arrays.asList(values));
+		}
+		return this;
+	}
+
+	/**
+	 * Associates the specified values with the key. If the key was already present
+	 * any existing values are removed and replaced with the new list.
+	 * null values are ignored. If all values are null, the put is ignored
+	 *
+	 * @param key the key
+	 * @param values the values
+	 * @throws IllegalArgumentException if the key is null or empty or contains invalid characters
+	 *         -or- if any value contains invalid characters
+	 */
+	public Headers put(String key, Collection<String> values) {
+		_put(key, values);
+		return this;
+	}
+
+	// the put delegate
+	private void _put(String key, Collection<String> values) {
+		if (values != null) {
+			checkKey(key);
+			List<String> checked = checkValues(values);
+			if (!checked.isEmpty()) {
+				headerMap.put(key, checked);
+				serialized = null; // since the data changed, clear this so it's rebuilt
+			}
+		}
+	}
+
+	/**
+	 * Removes each key and its values if the key was present
+	 *
+	 * @param keys the key or keys to remove
+	 */
+	public void remove(String... keys) {
+		for (String key : keys) {
+			headerMap.remove(key);
+		}
+		serialized = null; // since the data changed, clear this so it's rebuilt
+	}
+
+	/**
+	 * Removes each key and its values if the key was present
+	 *
+	 * @param keys the key or keys to remove
+	 */
+	public void remove(Collection<String> keys) {
+		for (String key : keys) {
+			headerMap.remove(key);
+		}
+		serialized = null; // since the data changed, clear this so it's rebuilt
+	}
+
+	public int size() {
+		return headerMap.size();
+	}
+
+	public boolean isEmpty() {
+		return headerMap.isEmpty();
+	}
+
+	public void clear() {
+		headerMap.clear();
+	}
+
+	public boolean containsKey(String key) {
+		return headerMap.containsKey(key);
+	}
+
+	public Set<String> keySet() {
+		return headerMap.keySet();
+	}
+
+	public List<String> values(String key) {
+		List<String> list = headerMap.get(key);
+		return list == null ? null : Collections.unmodifiableList(list);
+	}
+
+	public int serializedLength() {
+		return getSerialized().length;
+	}
+
+	public byte[] getSerialized() {
+		if (serialized == null) {
+			ByteArrayBuilder bab = new ByteArrayBuilder()
+					.append(VERSION_BYTES, VERSION_BYTES_LEN);
+			for (String key : headerMap.keySet()) {
+				for (String value : values(key)) {
+					bab.append(key);
+					bab.append(COLON_BYTES, COLON_BYTES_LEN);
+					bab.append(value);
+					bab.append(CRLF_BYTES, CRLF_BYTES_LEN);
```

Comment:
These append operations do not look efficient as they appear to introduce a lot of allocation overhead, you really should be pre-allocating the exact amount of bytes necessary rather than expanding the buffer repeatedly. This should be fairly straight forward if you replace the strings in `headerMap` with `ByteBuffer`'s.

Reply:
Could track size when adding/putting/removing headers and allocate the exact amount here. tbh, I'm not sure how much difference that'd make.  For maximum performance in an application you'd attempt to keep headers static and re-use messages (thus only one initial allocation and serialization).

Reply:
Take a look how I did it [here](https://github.com/nats-io/nats.java/commit/f833b99d51166821afcd544d29a0f32a84d94ace).

Reply:
I considered some tracking of the length of data added and removed, wasn't sure if it was worth the effort. We can add some optimization next round.

Reply:
You can just do it like [this](https://github.com/nats-io/nats.java/commit/92e4c6c9515c3e6ee728fd40b2605aff093ccb61#diff-5224933708c6d2f3fc83858e058e54cb3aff6e55ae4e679444ea05936c9e19a6R295-R304) which doesn't require serialization/de-serialization to compute the length.